### PR TITLE
Soft limit messaging

### PR DESCRIPTION
**Pull request by Naatan**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-1490" title="DX-1490" target="_blank">TaskDX-1490</a>  State tool displays a soft-limit message
  </td></table>
  <br />

### CODE REVIEW

File: internal/runbits/rtusage/rtusage.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/ActiveState/cli/internal/notify"
+	"github.com/ActiveState/cli/internal/output"
+	"strconv"
+)
+
+type dataHandler interface {
+	CheckRuntimeUsage(ctx context.Context, organizationName string) (*graph.CheckRuntimeUsageResponse, error)
+}
+
+type checkFunc func(ctx context.Context, organizationName string) (*graph.CheckRuntimeUsageResponse, error)
```
Comment (MDrakos) [inline internal/runbits/rtusage/rtusage.go:25]: This isn't used

File: internal/runbits/rtusage/rtusage.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+	usage, err := data.CheckRuntimeUsage(context.Background(), orgName)
+	if err != nil {
+		multilog.Error("Soft limit: Failed to check runtime usage in heartbeat handler: %s", errs.JoinMessage(err))
+		return
+	}
+
+	if usage.Usage > usage.Limit {
+		err := notify.Send(locale.Tl("runtime_limit_reached_title", "Runtime Limit Reached"),
+			locale.Tl("runtime_limit_reached_msg", "Heads up! You've reached your runtime limit for ActiveState-Labs."),
```
Comment (MDrakos) [inline internal/runbits/rtusage/rtusage.go]: I don't think `ActiveState-Labs` should be hard-coded here.

File: internal/runners/activate/activate.go [hunk 3]
```diff
@@ -2,6 +2,7 @@ package activate
 
 import (
 	"fmt"
+	"github.com/ActiveState/cli/internal/runbits/rtusage"
```
Comment (MDrakos) [inline internal/runners/activate/activate.go:5]: Nit: tooling should organize this import

File: internal/runners/protocol/protocol.go [hunk 4]
```diff
@@ -30,16 +29,23 @@ func New(prime primeable) *Protocol {
...
 	parsed, err := url.Parse(params.URL)
 	if err != nil {
 		return locale.WrapError(err, "err_protocol_parse", "Invailid URL provided: {{.V0}}", params.URL)
 	}
-	trimmedPath := strings.TrimLeft(parsed.Path, "/")
-	namespace, err := project.ParseNamespace(trimmedPath)
+
+	// Host=platform means we're trying to open a link to the platform
+	if parsed.Host == "platform" {
+		open.Run("https://platform.activestate.com" + parsed.Path)
```
Comment (MDrakos) [inline internal/runners/protocol/protocol.go:41]: Missing error check

File: internal/runners/activate/activate.go (hunk 3 above)
Comment (MDrakos) [reply internal/runners/activate/activate.go:5]: There are a lot of out-of-place imports in the PR...

File: pkg/platform/api/graphql/request/usage.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+func (p *usage) Query() string {
+	return `
+		query ($organization_id: uuid!, $week: date) {
+			  organizations_runtime_usage(limit: 1, order_by: [{week_of: desc}], where: {_and: [{organization_id: {_eq: $organization_id}}, {week_of: {_gte: $week}}]}) {
+					week_of
+					total_runtimes
+					static_runtimes_state_deploy
+					static_runtimes_next_gen_installer
+					organization_id
```
Comment (MDrakos) [inline pkg/platform/api/graphql/request/usage.go:33]: Is this key necessary in this query?

**Review by MDrakos — CHANGES_REQUESTED** [review]

File: cmd/state-svc/internal/resolver/resolver.go [hunk 6]
```diff
@@ -158,16 +173,34 @@ func (r *Resolver) AnalyticsEvent(_ context.Context, category, action string, _l
...
 	r.rtwatch.Watch(pid, exec, dims)
 
-	return &graph.RuntimeUsageResponse{Received: true}, nil
+	return &graph.ReportRuntimeUsageResponse{Received: true}, nil
+}
+
+func (r *Resolver) CheckRuntimeUsage(_ context.Context, organizationName string) (*graph.CheckRuntimeUsageResponse, error) {
+	logging.Debug("CheckRuntimeUsage resolver")
+
+	usage, err := r.usageChecker.Check(organizationName)
```
[the remark this answers is not in the document]
Comment (Naatan) [reply cmd/state-svc/internal/resolver/resolver.go:191]: The usageChecker is the one that communicates with the graphql endpoint, and that's the request we want to cache. We don't need to cache requests between state tool and state-svc as those are cheap.

File: internal/runbits/rtusage/rtusage.go (hunk 2 above)
Comment (Naatan) [reply internal/runbits/rtusage/rtusage.go]: Woops!

File: internal/runners/activate/activate.go (hunk 3 above)
Comment (Naatan) [reply internal/runners/activate/activate.go:5]: Hmm I've been developing using Goland on Windows.. Seems I may have a setting or plugin on mac that I'm missing. Good eye!

File: pkg/platform/api/graphql/request/usage.go (hunk 5 above)
Comment (Naatan) [reply pkg/platform/api/graphql/request/usage.go]: Fair point, I'll remove the unused ones.

**Review by MDrakos — APPROVED** [review]

**Pull request merged**